### ISSUE TITLE
Model online training performance optimization

### ISSUE THREAD

**Issue by ZhengHSI** [sgl-project/SpecForge#67]
When reproducing the llama4 online training script, I found that it takes a long time to train an epoch, and the inference of the target model takes a lot of time.（8*H20）

<img width="3320" height="1760" alt="Image" src="https://github.com/user-attachments/assets/7b7b6f90-0c44-47a0-ac18-24db4dbef33e" />

<img width="754" height="648" alt="Image" src="https://github.com/user-attachments/assets/0826cbc4-c43f-4b94-bec3-f04e27109bba" />

Is there any optimization here?

**Comment by therealnaveenkamal** [comment]
looks interesting, can I work on this?

**Comment by FlamingoPg** [comment]
> looks interesting, can I work on this?

Cool

**Comment by therealnaveenkamal** [comment]
@FlamingoPg / @ZhengHSI 

Are we not using `tp_size=8` for this execution? I am assuming the target model is replicated on all the GPUs.

**Comment by FlamingoPg** [comment]
> [@FlamingoPg](https://github.com/FlamingoPg) / [@ZhengHSI](https://github.com/ZhengHSI)
> 
> Are we not using `tp_size=8` for this execution? I am assuming the target model is replicated on all the GPUs.

Sure, this might be a good idea. But the base model is probably around 2B in size, so I'm not sure if tensor parallelism (TP) would bring improvement.

**Comment by yubofredwang** [comment]
Sharing some profiling results here, the memory usage during training for 1.22B draft model peaks at ~22GB. The majority of the memory spike is from softmax in attention's forward + backward.

<img width="1686" height="482" alt="Image" src="https://github.com/user-attachments/assets/ed108c3d-5045-4700-aafd-a43f5e9c5a02" />

#69 should help with the memory usage. We will prioritize flex attention work.

**Comment by therealnaveenkamal** [comment]
@yubofredwang oh interesting, do you think I should start with flexattention and then think of other optimizations?

**Comment by yubofredwang** [comment]
Close by #97
